00-programming-fundamentals/intro-to-regex.md

```diff
@@ -0,0 +1,133 @@
+| --- | --- |
+|  .  |  Matches any character |
+|  *  |  Matches if the preceeding token occurs zero or more times.  For example `/[0-9]*/` matches zero or more digits so "",  "123", and "5" would match. |  
+|  +  |  Matches if the preceeding token occurs one or more times.  For example `/[0-9]+/` matches one or more digits so both "1" and "345" would match.  |
+|  ?  |  Marks the preceeding token optional.  So `/example[0-9]?/` matches "example" followed by one or no digits, thus "example", "example1" and "example13" all match.  |
+
+
+Examples:  
+
+- `.+`  - Matches one or more of any characters.  
+-  `/[0-9]+/` - Matches one or more digits
+-  `/[A-Z]*/` - Matches zero or more capital letters.
+-  `/[a-z_][A-Za-z0-9]*/`  - Matches a lower case letter or underscore followed by zero or more alphanumeric characters.  
+
+## Escape characters
+
+There are lots of characters that have special meanings in a Regex (such as the + or * characters).  Just like Strings you can use the backslash character to select the exact character in the text.  It can also be used as a shortcut for common classes of characters.
+
+Some examples include:
+
+| Escape Character |  |  
+| --- | --- |
+| \n  |  newline character |
+| \s  |  Any whitespace character (space, tab, newline) |
+| \S  |  Any non-whitespace character |
+| \d  |  Any digit |
+| \D  |  Any non-digit |
+| \\., \\+, \\*  | The literal character following the backslash  |
+
+**Practice:** Try to write a regex for any amount of US currency, for example it should match `$3.25`, `$102.73`, and `$0.25`.  You can see an answer [here](https://gist.github.com/CheezItMan/09f422c21e5c30ac69cc6d64bb16c3a0).
+
+## Repetitions
+
+The * and + characters allow a token to be repeated, but often, such as for a zip code, you will want to limit a token to a specific number of repetitions.  For that you can use the curly braces ({}).  A number placed in the curly braces will indicate how many times the preceding token can be repeated.  So for example `/[abc]{3}/` will allow the letters a, b or c to be repeated three times, so "aaa" would match, as would "abc" as would "cab".
+
+A range of repetitions can also be repeated by using two parameters in the curly braces.  For example `/[abc]{3, 5}/` would allow the characters to repeat between 3 and 5 times.
+
+## Capture Groups
+
+We can also construct groups of characters which can be combined with special characters using the parentheses.  For example `/(ada *)+/` would match one or more "ada" lines separated by 0 or more spaces.  So "ada ada", "ada", and "ada ada   ada" would all match.
+
+So `/a(ab)*a/` would match the strings:  "aaba", "aa", "aababa" because it requires an a followed by zero or more "ab" blocks and ending with an "a".
+
+Similarly, `/\$(\d)+\.(\d){2}` would match us currency, like you were asked to deduce above requiring a dollar sign followed by one or more digits, followed by a period and ending with 2 digits.  
+
+
+
+### Practice
```

Comment:
I like the way this practice is called out in the text - could you update the one above to use this same header as well.

Reply:
Done